# Data Preprocessing - Filling NaN values
## Filling NaN values is a crucial step in data precprocessing as it gives Machine Learning models data to work with 
## In this notebook, we use the records with complete data to help fill in the NaN values using Linear Regression

### We start by importing the necessary libraries and loading the dataset into the notebook

In [131]:
df = pd.read_csv("F:\\School\\2-2\\Data Mining\\Project\\sub-division_rainfall_act_dep_1901-2015.csv",
                encoding = "unicode_escape")
df.head()

SUBDIVISION       YEAR                 Parameter    JAN  \
0  ANDAMAN & NICOBAR ISLANDS  1901-2015                      Mean   49.2   
1  ANDAMAN & NICOBAR ISLANDS  1901-2015        Standard deviation   71.3   
2  ANDAMAN & NICOBAR ISLANDS  1901-2015  Coefficient of variation  144.9   
3  ANDAMAN & NICOBAR ISLANDS       1901                    Actual   49.2   
4  ANDAMAN & NICOBAR ISLANDS       1901      Percentage departure   -0.1   

     FEB    MAR   APR    MAY    JUN    JUL    AUG    SEP    OCT    NOV    DEC  \
0   27.6   30.0  72.2  355.6  471.4  397.5  400.5  431.3  289.5  233.0  153.3   
1   38.8   43.6  66.8  151.2  147.0  151.9  142.6  146.7   99.4  119.4  129.5   
2  140.4  145.5  92.5   42.5   31.2   38.2   35.6   34.0   34.3   51.2   84.5   
3   87.1   29.2   2.3  528.8  517.5  365.1  481.1  332.6  388.5  558.2   33.6   
4  215.9   -2.6 -96.8   48.7    9.8   -8.2   20.1  -22.9   34.2  139.5  -78.1   

   ANNUAL     JF    MAM    JJAS    OND  
0  2911.0   76.8  457.8  1700.7  675.8  
1   395.9   81.1  176.8   286.3  199.0  
2    13.6  105.6   38.6    16.8   29.4  
3  3373.2  136.3  560.3  1696.3  980.3  
4    15.9   77.5   22.4    -0.3   45.1

In [132]:
df = df.drop(['ANNUAL', 'JF', 'MAM',
       'JJAS', 'OND'], axis=1)
df.head()

SUBDIVISION       YEAR                 Parameter    JAN  \
0  ANDAMAN & NICOBAR ISLANDS  1901-2015                      Mean   49.2   
1  ANDAMAN & NICOBAR ISLANDS  1901-2015        Standard deviation   71.3   
2  ANDAMAN & NICOBAR ISLANDS  1901-2015  Coefficient of variation  144.9   
3  ANDAMAN & NICOBAR ISLANDS       1901                    Actual   49.2   
4  ANDAMAN & NICOBAR ISLANDS       1901      Percentage departure   -0.1   

     FEB    MAR   APR    MAY    JUN    JUL    AUG    SEP    OCT    NOV    DEC  
0   27.6   30.0  72.2  355.6  471.4  397.5  400.5  431.3  289.5  233.0  153.3  
1   38.8   43.6  66.8  151.2  147.0  151.9  142.6  146.7   99.4  119.4  129.5  
2  140.4  145.5  92.5   42.5   31.2   38.2   35.6   34.0   34.3   51.2   84.5  
3   87.1   29.2   2.3  528.8  517.5  365.1  481.1  332.6  388.5  558.2   33.6  
4  215.9   -2.6 -96.8   48.7    9.8   -8.2   20.1  -22.9   34.2  139.5  -78.1

In [133]:
print(df.isnull().sum().sum())

297


We can see that the current dataset has 297 NaN values that we need to take care of

In [134]:
df_actual = pd.DataFrame(columns = ['SUBDIVISION', 'YEAR', 'Parameter', 'JAN', 'FEB', 'MAR', 'APR', 'MAY',
       'JUN', 'JUL', 'AUG', 'SEP', 'OCT', 'NOV', 'DEC'])
df_per_dep = df_actual
df_no_dis = df_actual

In [135]:
for i in range(0, df.shape[0]):
    if df.loc[i]["Parameter"] == "Actual":
        df_actual = df_actual.append(df.loc[i])
    elif df.loc[i]["Parameter"] == "Percentage departure":
        df_per_dep = df_per_dep.append(df.loc[i])
    elif df.loc[i]["Parameter"] == "No. of districts":
        df_no_dis = df_no_dis.append(df.loc[i])

We split the data frame into 3 categories accroding to their paramter - <br>
1) Actual <br>
2) Percentage Departure <br>
3) No. of districts <br>

In [136]:
print(df_actual.shape)
print(df_per_dep.shape)
print(df_per_dep.shape)

(4116, 15)
(4116, 15)
(4116, 15)


In [137]:
print(df_actual.isnull().sum().sum())
print(df_per_dep.isnull().sum().sum())
print(df_no_dis.isnull().sum().sum())
print(df.isnull().sum().sum())

70
227
0
297


We can see the break up on NaN values in each category <br>
We use the interpolate function of pandas to fill in these NaN values with the average of the year before and year after. <br>
Eg - if 1923 has an NaN value, it will be replaced with the average of 1922 and 1924

In [138]:
df_actual = df_actual.interpolate(method='linear', limit_direction='both')
df_per_dep = df_per_dep.interpolate(method='linear', limit_direction='both')
df_no_dis = df_no_dis.interpolate(method='linear', limit_direction='both')

In [139]:
print(df_actual.isnull().sum().sum())
print(df_per_dep.isnull().sum().sum())
print(df_no_dis.isnull().sum().sum())

0
0
0


As we can see, there are now no NaN values!

In [128]:
df_actual.to_csv("F:\\School\\2-2\\Data Mining\\Project\\rainfall_actual.csv")
df_per_dep.to_csv("F:\\School\\2-2\\Data Mining\\Project\\rainfall_per_dep.csv")
df_no_dis.to_csv("F:\\School\\2-2\\Data Mining\\Project\\rainfall_no_dis.csv")

We finish by saving these dataframes as csv workbooks so we can use them for further analysis